scipy/stats/_hypotests.py

```diff
@@ -1404,6 +1406,695 @@ def cramervonmises_2samp(x, y, method='auto'):
+
+    # (Of course, the user's statistic doesn't know what we've done here,
+    # so we need to pass it what it's expecting.)
+    def statistic_wrapped(*data, axis):
+        data = np.swapaxes(data, 0, -1)
+        if n_samples == 1:
+            data = data[0:1]
+        return statistic(*data, axis=axis)
+
+    return _calculate_null_pairings(data, statistic_wrapped, n_permutations,
+                                    batch, random_state)
+
+
+def _permutation_test_iv(data, statistic, permutation_type, vectorized,
+                         permutations, batch, alternative, axis, random_state):
+    """Input validation for `permutation_test`."""
+
+    axis_int = int(axis)
+    if axis != axis_int:
+        raise ValueError("`axis` must be an integer.")
+
+    permutation_types = {'samples', 'pairings', 'both'}
+    permutation_type = permutation_type.lower()
+    if permutation_type not in permutation_types:
+        raise ValueError(f"`permutation_type` must be in {permutation_types}.")
+
+    if vectorized not in {True, False}:
+        raise ValueError("`vectorized` must be `True` or `False`.")
+
+    message = "`data` must be a tuple containing at least two samples"
+    try:
+        if len(data) < 2 and permutation_type == "both":
+            raise ValueError(message)
+    except TypeError:
+        raise TypeError(message)
+
+    data = _broadcast_arrays(data, axis)
+    data_iv = []
+    for sample in data:
+        sample = np.atleast_1d(sample)
+        if sample.shape[axis] <= 1:
+            raise ValueError("each sample in `data` must contain two or more "
+                             "observations along `axis`.")
+        sample = np.moveaxis(sample, axis_int, -1)
+        data_iv.append(sample)
+
+    permutations_int = (int(permutations) if not np.isinf(permutations)
+                        else np.inf)
+    if permutations != permutations_int or permutations_int <= 0:
+        raise ValueError("`permutations` must be a positive integer.")
+
+    if batch is None:
+        batch_iv = batch
+    else:
+        batch_iv = int(batch)
+        if batch != batch_iv or batch_iv <= 0:
+            raise ValueError("`batch` must be a positive integer or None.")
+
+    alternatives = {'two-sided', 'greater', 'less'}
+    alternative = alternative.lower()
+    if alternative not in alternatives:
+        raise ValueError(f"`alternative` must be in {alternatives}")
+
+    random_state = check_random_state(random_state)
+
+    return (data_iv, statistic, permutation_type, vectorized, permutations_int,
+            batch_iv, alternative, axis_int, random_state)
+
+
+# This is mostly to protect me while developing. Eventually we want to batch
+# the computations, so this will be less of an issue, but if it sounds like a
+# good feature, let me know.
+def _memory_check(data, permutations):
+    needed = sum([sample.size for sample in data]) * permutations * 8
+    available = 1e9  # psutil.virtual_memory()[1] # can't use in SciPy
```

Comment:
i was not sure about this check, can you explain?

Reply:
The idea is to add some protection against crashes due to consuming all available memory. I hard-coded a number that was reasonable for my machine because the library I found for checking available memory can't be assumed as a dependency. I didn't look around for other options, but I think for now I'll just remove the check.

Reply:
Removed.